File: src/test/java/sirius/kernel/commons/TrieTest.java
package sirius.kernel.commons;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;

import static org.junit.jupiter.api.Assertions.*;

class TrieTest {

    private static Trie<Integer> trie;

    @BeforeAll
    static void createTrie() {
        trie = Trie.create();
        trie.put("one", 1);
        trie.put("on", 2);
        trie.put("one1", 3);
        trie.put("two", 4);
        trie.put("three", 5);
        trie.put("thrae", 6);
        trie.put("th", 7);
    }

    @Test
    void isFilled() {
        String check = "I'd like to have three beer please";
        Trie.ContainmentIterator<Integer> iter = trie.iterator();

        int found = 0;

        for (int i = 0; i < check.length(); i++) {
            if (!iter.doContinue(check.charAt(i))) {
                if (iter.isCompleted()) {
                    found = iter.getValue();
                }
                iter.resetWith(check.charAt(i));
            }
        }

        assertEquals(5, found);

        assertEquals(2, (int) trie.get("on"));
        assertNull(trie.get("onx"));
        assertTrue(trie.containsKey("thrae"));
        assertFalse(trie.containsKey("thre"));
    }

    @Test
    void keySet() {
        assertEquals(7, trie.size());
        assertEquals(new HashSet<>(Arrays.asList("one", "on", "one1", "two", "three", "thrae", "th")), trie.keySet());
        assertEquals(new HashSet<>(Arrays.asList("one", "on", "one1", "two", "three", "thrae", "th")),
                     trie.getAllKeysBeginningWith(""));
        assertEquals(new HashSet<>(Arrays.asList("one", "on", "one1")), trie.getAllKeysBeginningWith("on"));
        assertEquals(Collections.singleton("three"), trie.getAllKeysBeginningWith("three"));
        assertEquals(0, trie.getAllKeysBeginningWith("threee").size());
    }
}

File: src/test/kotlin/sirius/kernel/commons/TrieTest.kt
/*
 * Made with all the love in the world
 * by scireum in Remshalden, Germany
 *
 * Copyright by scireum GmbH
 * http://www.scireum.de - info@scireum.de
 */

package sirius.kernel.commons

import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Tests the [Trie] class.
 */
class TrieTest {
    @Test
    fun isFilled() {
        val check = "I'd like to have three beer please"
        val iter = trie.iterator()
        var found = 0
        for (index in check.indices) {
            if (!iter.doContinue(check[index].code)) {
                if (iter.isCompleted()) {
                    found = iter.getValue()
                }
                iter.resetWith(check[index].code)
            }
        }
        assertEquals(5, found)
        assertEquals(2, trie["on"] as Int)
        assertNull(trie["onx"])
        assertTrue { trie.containsKey("thrae") }
        assertFalse { trie.containsKey("thre") }
    }

    @Test
    fun keySet() {
        assertEquals(7, trie.size())
        assertEquals(setOf("one", "on", "one1", "two", "three", "thrae", "th"), trie.keySet())
        assertEquals(setOf("one", "on", "one1", "two", "three", "thrae", "th"), trie.getAllKeysBeginningWith(""))
        assertEquals(setOf("one", "on", "one1"), trie.getAllKeysBeginningWith("on"))
        assertEquals(setOf("three"), trie.getAllKeysBeginningWith("three"))
        assertEquals(0, trie.getAllKeysBeginningWith("threee").size)
    }

    companion object {
        private lateinit var trie: Trie<Int>

        @JvmStatic
        @BeforeAll
        fun createTrie() {
            trie = Trie.create()
            trie.put("one", 1)
            trie.put("on", 2)
            trie.put("one1", 3)
            trie.put("two", 4)
            trie.put("three", 5)
            trie.put("thrae", 6)
            trie.put("th", 7)
        }
    }
}
